presto-main/src/main/java/com/facebook/presto/operator/aggregation/OptimizedTypedSet.java

```diff
@@ -0,0 +1,339 @@
+
+        this.elementType = requireNonNull(elementType, "elementType must not be null");
+        this.hashCapacity = arraySize(maxPositionCount, FILL_RATIO);
+        this.hashMask = hashCapacity - 1;
+
+        blocks = new Block[expectedBlockCount];
+        positionsForBlocks = new ArrayList<>(expectedBlockCount);
+        blockPositionByHash = initializeHashTable();
+    }
+
+    /**
+     * Union the set by adding the elements of the block, eliminating duplicates.
+     */
+    public void union(Block block)
+    {
+        currentBlockIndex++;
+        ensureBlocksCapacity(currentBlockIndex + 1);
+        blocks[currentBlockIndex] = block;
+
+        int positionCount = block.getPositionCount();
+        int[] positions = new int[positionCount];
+
+        // Add the elements to the hash table. Since union can only increase the set size, there is no need to create a separate hashtable.
+        int positionsIndex = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int hashPosition = getInsertPosition(blockPositionByHash, getMaskedHash(hashPosition(elementType, block, i)), block, i);
+            if (hashPosition != INVALID_POSITION) {
+                // There is no need to test if adding element is successful since it's on the same hash table
+                addElement(blockPositionByHash, hashPosition, block, i);
+                positions[positionsIndex++] = i;
+            }
+        }
+
+        getPositionsForBlocks().add(positionsList(positions, 0, positionsIndex));
+        size += positionsIndex;
+    }
+
+    /**
+     * Intersect the set with the elements of the block, eliminating duplicates.
+     */
+    public void intersect(Block block)
+    {
+        currentBlockIndex++;
+        ensureBlocksCapacity(currentBlockIndex + 1);
+        blocks[currentBlockIndex] = block;
+
+        if (currentBlockIndex == 0) {
+            // This set was an empty set, so the result set should also be an empty set.
+            positionsForBlocks.add(EMPTY_SELECTED_POSITIONS);
+            return;
+        }
+
+        int positionCount = block.getPositionCount();
+        int[] positions = ensureCapacity(positionsForBlocks.get(currentBlockIndex - 1).getPositions(), positionCount);
+
+        // We need to create a new hash table because the elements in the set may be removed
+        long[] newBlockPositionByHash = initializeHashTable();
+
+        int positionsIndex = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int hash = getMaskedHash(hashPosition(elementType, block, i));
+            int positionInBlockPositionByHash = getInsertPosition(blockPositionByHash, hash, block, i);
+            if (positionInBlockPositionByHash == INVALID_POSITION) {
+                // add to the hash table if it exists in blockPositionByHash
+                if (addElement(newBlockPositionByHash, hash, block, i)) {
+                    positions[positionsIndex++] = i;
+                }
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 127:136, 168:178.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.